runtime/vm/createramclass.cpp

```diff
@@ -1139,116 +1141,137 @@ fillJITVTableSlot(J9VMThread *vmStruct, UDATA *currentSlot, J9Method *currentMet
-										J9UTF8 *superclassVTableMethodNameUTF = J9ROMMETHOD_NAME(superclassVTableROMMethod);
-										errorData->loader1 = classLoader;
-										errorData->class1NameUTF = newClassNameUTF;
-										errorData->loader2 = superclassVTableMethodLoader;
-										errorData->class2NameUTF = superclassVTableMethodClassNameUTF;
-										errorData->exceptionClassNameUTF = superclassVTableMethodClassNameUTF;
-										errorData->methodNameUTF = superclassVTableMethodNameUTF;
-										errorData->methodSigUTF = superclassVTableMethodSigUTF;
-										vTableWriteIndex = 0;
-										goto done;
-									}
 
-								}
-								vTableAddress[superclassVTableIndex] = (UDATA)storeValue;
+				/* Look at the vTable of each superclass, not just the immediate one */
+				J9Class *currentSuperclass = superclass;
+				do {
+					UDATA *currentSuperclassVTable = (UDATA *)(currentSuperclass + 1);
+					/* Stop the search if the superclass does not contain an entry at the search index */
+					if (currentSuperclassVTable[0] < superclassVTableIndex) {
+						break;
+					}
+					J9Method *currentSuperclassVTableMethod = (J9Method *)currentSuperclassVTable[superclassVTableIndex];
+					J9Class *currentSuperclassVTableMethodClass = J9_CLASS_FROM_METHOD(currentSuperclassVTableMethod);
+					J9ROMMethod *currentSuperclassVTableROMMethod = J9_ROM_METHOD_FROM_RAM_METHOD(currentSuperclassVTableMethod);
+					UDATA currentSuperclassModifiers = currentSuperclassVTableROMMethod->modifiers;
+					/* Private methods are not stored in the vTable.
+					 * Public and protected are always overridden.
+					 * Default is overridden only by a method in the same package.
+					 */
+					if (J9_ARE_ANY_BITS_SET(currentSuperclassModifiers, J9_JAVA_PROTECTED | J9_JAVA_PUBLIC)
+						|| (currentSuperclassVTableMethodClass->packageID == localPackageID)
+					) {
+						overridden = TRUE;
+						newSlotRequired = FALSE;
+						break;
+					}
+					/* For class file versions below 51, check only the immediate superclass */
+					if (romClass->majorVersion < 51) {
+						break;
+					}
+					currentSuperclass = VM_VMHelpers::getSuperclass(currentSuperclass);
+				} while(NULL != currentSuperclass);
+				if (overridden) {
+					if ((((UDATA)storeValue & VTABLE_SLOT_TAG_MASK) == ROM_METHOD_ID_TAG)
+						|| (vTableAddress[superclassVTableIndex] == (UDATA)superclassVTableMethod)
+					) {
+						if ((modifiers & J9_JAVA_FINAL) == J9_JAVA_FINAL) {
+							vmThread->tempSlot = (UDATA)romMethod;
+						}
+						/* fill in vtable, override parent slot */
+						J9ClassLoader *superclassVTableMethodLoader = superclassVTableMethodClass->classLoader;
+						if (superclassVTableMethodLoader != classLoader) {
+							J9UTF8 *superclassVTableMethodSigUTF = J9ROMMETHOD_SIGNATURE(superclassVTableROMMethod);
+							if (0 != j9bcv_checkClassLoadingConstraintsForSignature(vmThread, classLoader, superclassVTableMethodLoader, sigUTF, superclassVTableMethodSigUTF)) {
```

Comment:
The CL constraint check is correct because the vtables are built on top of the parent's vtable - we'd have already constrained the parent loader if required.  This pass through is only required to constraint the current loader to the parent loader as the check is transitive.

Reply:
Not sure what you're complaining about here.

Reply:
Not complaining - convincing myself that the clconstraint check is correct while giving you an opportunity to agree or disagree.

Reply:
Any superclass can be used for the check, as long as the signature data comes from the matching class loader, which is the case here.